includes/class-sensei-notices.php

```diff
@@ -94,31 +86,41 @@ public function add_notice( string $content, string $type = 'alert', string $key
 				'type'    => $type,
 			];
 		}
+	}
 
-		// if a notice is added after we've printed, print it immediately.
-		if ( $this->has_printed ) {
-			$this->maybe_print_notices();
-		}
+	/**
+	 * Output the container to put the notices.
+	 *
+	 * @return void
+	 */
+	public function maybe_print_notices_container() {
+		?>
+		<div class="sensei-notices-container"></div>
+		<?php
 	}
 
 	/**
-	 * Output all notices added
+	 * Output all notices added as scripts to the last "sensei-notices-container" element.
 	 *
 	 * @return void
 	 */
-	public function maybe_print_notices() {
+	public function maybe_print_notices_on_script() {
 		if ( ! empty( $this->notices ) ) {
-			foreach ( $this->notices  as  $notice ) {
-				// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped -- Output escaped in generate_notice
-				echo $this->generate_notice( $notice['type'], $notice['content'] );
+			$result = '';
+			foreach ( $this->notices as $notice ) {
+				$result .= $this->generate_notice( $notice['type'], $notice['content'] );
 			}
+			?>
+			<script type="text/javascript">
+				( function(d) {
+					const containers = Array.from( d.querySelectorAll( ".sensei-notices-container" ) );
+					containers.pop().innerHTML = <?php echo wp_json_encode( $result ); ?>;
+				} )(document);
+			</script>
```

Comment:
Maybe this should use [`wp_add_inline_script`](https://developer.wordpress.org/reference/functions/wp_add_inline_script/) instead, but I'm not 100% sure.